### ISSUE TITLE
Incorrect syntax error

### ISSUE THREAD

**Issue by bjorn3** [rust-analyzer/rust-analyzer#3512]
```rust
&mut { [] }[mid..];
//         ^
// Syntax Error: expected SEMI
// expected one of `.`, `;`, `?`, `}`, or an operator, found `[`
// expected one of `.`, `;`, `?`, `}`, or an operator
```

**Comment by matklad** [comment]
Hm, I'd say this works as expected? Our error message ("expected SEMI") is not exactly ideal, but don't generally produce nice syntax error messages at the moment at all .

**Comment by bjorn3** [comment]
This is actually correct syntax. I believe I found this in libcore.

**Comment by matklad** [comment]
Hm, it gives syntax error to me: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f5a00ddfcd0f13aa1021053a30d56d15

**Comment by bjorn3** [comment]
I missed the `let v =` prefix. The original can be found at https://github.com/rust-lang/rust/blob/04e69e4f4234beb4f12cc76dcc53e2cc4247a9be/src/libcore/slice/sort.rs#L658

**Comment by matklad** [comment]
Ok, see the bug now, thanks for the clarification!
